src/features/element-hiding.js

```diff
@@ -1,78 +1,234 @@
-    const frameElements = [...node.querySelectorAll('iframe')]
+    // no sense wasting cycles checking if the page's body element is empty
+    if (node.tagName === 'BODY') {
+        return false
+    }
+    // use a DOMParser to remove all metadata elements before checking if
+    // the node is empty.
+    const parsedNode = parser.parseFromString(node.outerHTML, 'text/html').documentElement
+    parsedNode.querySelectorAll('base,link,meta,script,style,template,title,desc').forEach((el) => {
+        el.remove()
+    })
+
+    const visibleText = parsedNode.innerText.trim().toLocaleLowerCase().replace(/:$/, '')
+    const mediaContent = parsedNode.querySelector('video,canvas,picture')
+    const frameElements = [...parsedNode.querySelectorAll('iframe')]
     // about:blank iframes don't count as content, return true if:
     // - node doesn't contain any iframes
     // - node contains iframes, all of which are hidden or have src='about:blank'
     const noFramesWithContent = frameElements.every((frame) => {
         return (frame.hidden || frame.src === 'about:blank')
     })
+
     if ((visibleText === '' || adLabelStrings.includes(visibleText)) &&
         noFramesWithContent && mediaContent === null) {
         return true
     }
     return false
 }
 
-function hideMatchingDomNodes (rules) {
+/**
+ * Apply relevant hiding rules to page at set intervals
+ * @param {Object[]} rules
+ * @param {string} rules[].selector
+ * @param {string} rules[].type
+ */
+function applyRules (rules) {
+    // several passes are made to hide & unhide elements. this is necessary because we're not using
+    // a mutation observer but we want to hide/unhide elements as soon as possible, and ads
+    // frequently take from several hundred milliseconds to several seconds to load
+    // check at 0ms, 100ms, 200ms, 300ms, 400ms, 500ms, 1000ms, 1500ms, 2000ms, 2500ms, 3000ms
+    hideAdNodes(rules)
+    let immediateHideIterations = 0
+    const immediateHideInterval = setInterval(function () {
+        immediateHideIterations += 1
+        if (immediateHideIterations === 4) {
+            clearInterval(immediateHideInterval)
+        }
+        hideAdNodes(rules)
+    }, 100)
+
+    let delayedHideIterations = 0
+    const delayedHideInterval = setInterval(function () {
+        delayedHideIterations += 1
+        if (delayedHideIterations === 4) {
+            clearInterval(delayedHideInterval)
+        }
+        hideAdNodes(rules)
+    }, 500)
+
+    // check previously hidden ad elements for contents, unhide if content has loaded after hiding.
+    // we do this in order to display non-tracking ads that aren't blocked at the request level
+    // check at 750ms, 1500ms, 2250ms, 3000ms
+    let unhideIterations = 0
+    const unhideInterval = setInterval(function () {
+        unhideIterations += 1
+        if (unhideIterations === 3) {
+            clearInterval(unhideInterval)
+        }
+        unhideLoadedAds()
+    }, 750)
```

Comment:
After these intervals can we clear all of out caches `hiddenElements` and `appliedRules`?

Reply:
Good call, I'll address.